### PR TITLE
fix: remove deps topic listener during service close

### PR DESCRIPTION
**Pull request by alter-mage**
**Issue #, if available:**

**Description of changes:**
Remove the service dependency topic watcher when closing the service. On a high level, while Nucleus is performing bootstrap and requesting restarts/reboots when required, removing the watcher should prevent new services from requesting installation, causing their lifecycle steps to be executed and possibly interrupted by restarts/reboots.

**Why is this change necessary:**
When device receives a deployment with bootstrap steps, the deployment is activated via a KernelUpdateActivator which stops all services, applies bootstrap, requests nucleus restart/device reboot, requests new services to be started.
Nucleus collects the new config to be merged as a map of config keys and objects and closes all services before applying the merge. The config merge invokes various subscribers to react to the change in the config.
We let inbuilt java methods take care of the ordering in which service config needs to be updated
https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/e447acd4e0946a1c188d5bf4876de626a727cf28/src/main/java/com/aws/greengrass/config/Topics.java#L316-L320

The service `main` is dependent on all the new services for each deployment and registers a dependency topic listener whenever the new config for `main` is being merged
https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/e447acd4e0946a1c188d5bf4876de626a727cf28/src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java#L199-L210

Any change in the config path `services/<service-name>/dependencies` triggers a subscription which loads all of `service-name` (mostly `main`) dependency services.
If a dependency service config is updated after `main` loads that service then a lifecycle change of that service will trigger a re-installation and cause its lifecycle steps to start executing
https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/e447acd4e0946a1c188d5bf4876de626a727cf28/src/main/java/com/aws/greengrass/lifecyclemanager/GenericExternalService.java#L129-L136C3

This is not a desirable experience as services should only be started once all bootstrap steps have been executed and Nucleus explicitly asks for the services to be started.

**How was this change tested:**
- [x] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [x] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java [hunk 1]
```diff
@@ -449,6 +449,7 @@ protected CompletableFuture<Void> close(boolean waitForDependers) {
                         logger.error("Interrupted waiting for dependers to exit");
                     }
                 }
+                externalDependenciesTopic.remove();
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java:452]: As communicated offline, directly calling `externalDependenciesTopic.remove()` will remove the entire node. You should just remove the watcher

File: src/main/java/com/aws/greengrass/config/Topic.java [hunk 2]
```diff
@@ -53,6 +54,27 @@ public Topic subscribe(Subscriber s) {
...
+     * This is the preferred way to get a value from a configuration. Instead of {@code setValue(configValue.getOnce())}
+     * use {@code configValue.get((nv,ov)->setValue(nv)) }
+     * This way, every change to the config file will get forwarded to the object.
+     *</p>
+     *
+     * @param s subscriber
+     * @param service service subscribing to dependency topic
+     * @return this topic
+     */
+    public Topic subscribe(Subscriber s, GreengrassService service) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/config/Topic.java:69]: why are you adding this? This is highly specific logic for services which should not exist in the config.

Comment (shaguptashaikh) [reply src/main/java/com/aws/greengrass/config/Topic.java:69]: This is oddly placed in the Topic class, should be in GreengrassService

File: src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java [hunk 3]
```diff
@@ -17,14 +19,12 @@
 import com.aws.greengrass.integrationtests.BaseITCase;
 import com.aws.greengrass.integrationtests.ipc.IPCTestUtils;
 import com.aws.greengrass.integrationtests.util.ConfigPlatformResolver;
-import com.aws.greengrass.lifecyclemanager.GenericExternalService;
-import com.aws.greengrass.lifecyclemanager.GlobalStateChangeListener;
-import com.aws.greengrass.lifecyclemanager.GreengrassService;
-import com.aws.greengrass.lifecyclemanager.Kernel;
+import com.aws.greengrass.lifecyclemanager.*;
```
Comment (shaguptashaikh) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java:22]: No wildcard imports, undo this change and import specific additional things you need

File: src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java [hunk 4]
```diff
@@ -217,6 +234,45 @@ void GIVEN_kernel_running_single_service_WHEN_merge_changes_service_THEN_service
         }
     }
 
+    @Test
+    void GIVEN_kernel_running_single_service_WHEN_merge_config_with_bootstrap_THEN_dependency_watcher_is_removed()
```
Comment (shaguptashaikh) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java:238]: Why were we not able to make a test for the kernel update workflow and have to try the regular deployment merge?

File: src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java (hunk 3 above)
Comment (alter-mage) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java:22]: yep!

File: src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java (hunk 4 above)
Comment (shaguptashaikh) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java:238]: Ok, looks like you are trying to merge something with bootstrap so it should take the bootstrap route, but what is this test doing? I don't see any verifications after the merge is performed

File: src/main/java/com/aws/greengrass/config/Topic.java (hunk 2 above)
Comment (alter-mage) [reply src/main/java/com/aws/greengrass/config/Topic.java:69]: You mean the new subscribe method or the topic watchers?

File: src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java [hunk 5]
```diff
@@ -84,6 +86,8 @@ public class GreengrassService implements InjectionActions {
 
     // dependencies that are explicitly declared by customer in config store.
     private final Topic externalDependenciesTopic;
+    @Getter @Setter
+    private Watcher externalDependenciesTopicWatcher;
```
Comment (shaguptashaikh) [inline src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java:90]: Will there always be just one watcher?

File: src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java (hunk 4 above)
Comment (alter-mage) [reply src/integrationtests/java/com/aws/greengrass/integrationtests/deployment/DeploymentConfigMergingTest.java:238]: I was using this test to investigate the root cause of this problem. Since KernelUpdateActivator eventually calls System.exit, I dont think I can use this integ test for any verification and planning to remove this. The plan is to instead test the E2E with a UAT. But I have kept the test here for now as a draft.

File: src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java (hunk 5 above)
Comment (alter-mage) [reply src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java:90]: There could be more, but removing just the one watcher which is registered when the service is initialized should be sufficient

File: src/main/java/com/aws/greengrass/config/Topic.java (hunk 2 above)
Comment (shaguptashaikh) [reply src/main/java/com/aws/greengrass/config/Topic.java:69]: Both

Comment (shaguptashaikh) [reply src/main/java/com/aws/greengrass/config/Topic.java:69]: Why does this need to be done this way? Why can this code not live in the service class?

File: src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java (hunk 5 above)
Comment (shaguptashaikh) [reply src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java:90]: There could be more watchers for dependencies? and if so, why is removing just one sufficient?

File: src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java [hunk 6]
```diff
@@ -116,4 +131,35 @@ void GIVEN_service_WHEN_dependencies_change_THEN_service_dependencies_updates()
...
+        serviceConfiguration.read(configPath);
+        Topics root = serviceConfiguration.getRoot();
+        Topics serviceTopics = spy(root.findTopics(SERVICES_NAMESPACE_TOPIC, "B"));
+        Topic externalDependencyTopic = mock(Topic.class);
+
+        doReturn(context).when(serviceTopics).getContext();
+        doReturn(externalDependencyTopic).when(serviceTopics).createLeafChild(eq(SERVICE_DEPENDENCIES_NAMESPACE_TOPIC));
+        doReturn(externalDependencyTopic).when(externalDependencyTopic).dflt(any());
+        doReturn(Collections.emptyList()).when(externalDependencyTopic).getOnce();
+        lenient().doReturn(externalDependencyTopic).when(externalDependencyTopic).subscribe(any());
```
Comment (shaguptashaikh) [inline src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:149]: Why lenient? if this is specifically being mocked for this test, then it either gets invoked or it doesn't

Comment (alter-mage) [reply src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:149]: I always run into a `Strict stubbing argument mismatch` if I do not use lenient

Comment (alter-mage) [reply src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:149]: Okay, I dont think this is required at all

File: src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java (hunk 5 above)
Comment (alter-mage) [reply src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java:90]: I have not come across a situation where there were more than one watchers for a service dependency topic, I did not exhaustively look at situations on this. However, it seemed like the watcher which was registered during the initial watcher registration still being alive when services were closed was causing all the problems. I also wanted to make minimal changes on this.

Comment (alter-mage) [reply src/main/java/com/aws/greengrass/lifecyclemanager/GreengrassService.java:90]: But if there is a recommendation that the entire Watchers set should be emptied when a service is being closed then that can be done as well

File: src/main/java/com/aws/greengrass/config/Topic.java (hunk 2 above)
Comment (alter-mage) [reply src/main/java/com/aws/greengrass/config/Topic.java:69]: I wanted to maintain the original definition of `.subscribe` while also allowing `GreengrassService` to store a reference of the dependency watcher

File: src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java [hunk 7]
```diff
@@ -116,4 +131,35 @@ void GIVEN_service_WHEN_dependencies_change_THEN_service_dependencies_updates()
...
+        serviceConfiguration.read(configPath);
+        Topics root = serviceConfiguration.getRoot();
+        Topics serviceTopics = spy(root.findTopics(SERVICES_NAMESPACE_TOPIC, "B"));
+        Topic externalDependencyTopic = mock(Topic.class);
+
+        doReturn(context).when(serviceTopics).getContext();
+        doReturn(externalDependencyTopic).when(serviceTopics).createLeafChild(eq(SERVICE_DEPENDENCIES_NAMESPACE_TOPIC));
+        doReturn(externalDependencyTopic).when(externalDependencyTopic).dflt(any());
+        doReturn(Collections.emptyList()).when(externalDependencyTopic).getOnce();
+//        doReturn(externalDependencyTopic).when(externalDependencyTopic).subscribe(any());
```
[the remark this answers is not in the document]
Comment (alter-mage) [reply src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:149]: I meant to delete that :D

**Review by indougnito — DISMISSED** [review]

File: src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java [hunk 8]
```diff
@@ -57,6 +70,8 @@ void beforeEach() throws IOException, URISyntaxException, ServiceLoadException {
         Path configPath = Paths.get(this.getClass().getResource("services.yaml").toURI());
         context = spy(new Context());
         context.put(Kernel.class, kernel);
+        ExecutorService executorService = Executors.newCachedThreadPool();
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:73]: do not create a new pool.

If you must, then make sure it gets shutdown.

File: src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java [hunk 9]
```diff
@@ -74,6 +90,7 @@ void beforeEach() throws IOException, URISyntaxException, ServiceLoadException {
     @AfterEach
     void afterEach() throws IOException {
         context.close();
+        executorService.shutdown();
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:93]: shutdownnow

Comment (alter-mage) [reply src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:93]: Got it, I used shutdownNow before but modified to shutdown
Is there a reason to prefer shutdownNow over shutdown?

Comment (MikeDombo) [reply src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java]: shutdown now kills anything that's currently running, we don't want anything stray to keep running.

**Review by indougnito — APPROVED** [review]

File: src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java [hunk 10]
```diff
@@ -116,4 +133,34 @@ void GIVEN_service_WHEN_dependencies_change_THEN_service_dependencies_updates()
...
+        AtomicBoolean watcherRemoved = new AtomicBoolean(false);
+        doAnswer((Answer<Void>) invocationOnMock -> {
+            watcherRemoved.set(true);
+            return null;
+        }).when(externalDependencyTopic).remove(any());
+
+        service.close().get(30, TimeUnit.SECONDS);
+
+        //THEN
+        assertTrue(watcherRemoved.get());
```
Comment (shaguptashaikh) [inline src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:164]: Does `service.close()` ever get undone expecting the service lifecycle for the same service instance to progress as usual? Or does it always only happen during nucleus shutdown or service removal?

Comment (shaguptashaikh) [reply src/test/java/com/aws/greengrass/lifecyclemanager/GreengrassServiceTest.java:164]: The reason I'm asking is to see whether something needs to be tested for that

**Review by shaguptashaikh — APPROVED** [review]

**Review by junfuchen99 — APPROVED** [review]
Nice work getting this across the finish line
